### PR TITLE
main.cpp fixes, refactoring

### PR DESCRIPTION
**Pull request by anzz1**
moved fixes and refactorings from #555 

- main: entering empty line passes back control without new input in interactive/instruct modes
- instruct mode: keep prompt fix
- instruct mode: duplicate instruct prompt fix
- refactor: move common console code from main->common

### CODE REVIEW

File: examples/common.cpp [hunk 1]
```diff
@@ -256,3 +265,47 @@ std::vector<llama_token> llama_tokenize(struct llama_context * ctx, const std::s
 
     return res;
 }
+
+/* Keep track of current color of output, and emit ANSI code if it changes. */
+void set_console_color(console_state* con_st, console_color_t color) {
```
Comment (slaren) [inline examples/common.cpp:270]: Consider replacing `con_st` with a reference to make it clear that NULL is not an acceptable value

File: examples/main/main.cpp [hunk 2]
```diff
@@ -218,24 +146,23 @@ int main(int argc, char ** argv) {
         return 1;
     }
 
-    params.n_keep    = std::min(params.n_keep,    (int) embd_inp.size());
+    // number of tokens to keep when resetting context
+    if (params.n_keep < 0 || params.n_keep > (int)embd_inp.size() || params.instruct) {
+        params.n_keep = (int)embd_inp.size();
+    }
```
Comment (slaren) [inline examples/main/main.cpp:152]: Update `gpt_print_usage` to explain that `--keep -1` will use the entire prompt automatically (and also show what is the default, which is currently missing)

Comment (anzz1) [reply examples/main/main.cpp:152]: :+1:

**Review by slaren — APPROVED** [review]
Looks good!

**Pull request merged**